File: 351004/Lazuta/services/publisher/src/services/EditorService.cpp
#include "EditorService.h"
#include <storage/database/EditorRepository.h>
#include <mapping/DtoMapper.h>
#include <exceptions/DatabaseException.h>
#include <exceptions/NotFoundException.h>
#include <exceptions/ValidationException.h>

namespace publisher
{

using namespace drogon_model::distcomp;
using namespace publisher::dto;

EditorService::EditorService(std::shared_ptr<EditorRepository> storage)
    : m_dao(storage)
{
}

EditorResponseTo EditorService::Create(const EditorRequestTo& request)
{
    request.validate();
    
    TblEditor entity = DtoMapper::ToEntity(request);
    auto result = m_dao->Create(entity);
    
    if (std::holds_alternative<DatabaseError>(result))
    {
        DatabaseError error = std::get<DatabaseError>(result);
        switch (error)
        {
            case DatabaseError::AlreadyExists:
                throw ValidationException("Editor with this login already exists");
            default:
                throw DatabaseException("Failed to create editor");
        }
    }
    
    int64_t id = std::get<int64_t>(result);
    auto getResult = m_dao->GetByID(id);
    
    if (std::holds_alternative<DatabaseError>(getResult))
    {
        throw DatabaseException("Failed to retrieve created editor");
    }
    
    return DtoMapper::ToResponse(std::get<TblEditor>(getResult));
}

EditorResponseTo EditorService::Read(int64_t id)
{
    auto result = m_dao->GetByID(id);
    
    if (std::holds_alternative<DatabaseError>(result))
    {
        DatabaseError error = std::get<DatabaseError>(result);
        if (error == DatabaseError::NotFound)
        {
            throw NotFoundException("Editor not found");
        }
        throw DatabaseException("Failed to retrieve editor");
    }
    
    return DtoMapper::ToResponse(std::get<TblEditor>(result));
}

EditorResponseTo EditorService::Update(const EditorRequestTo& request, int64_t id)
{
    request.validate();
    
    TblEditor entity = DtoMapper::ToEntityForUpdate(request, id);
    auto updateResult = m_dao->Update(id, entity);
    
    if (std::holds_alternative<DatabaseError>(updateResult))
    {
        DatabaseError error = std::get<DatabaseError>(updateResult);
        if (error == DatabaseError::NotFound)
        {
            throw NotFoundException("Editor not found for update");
        }
        throw DatabaseException("Failed to update editor");
    }
    
    auto getResult = m_dao->GetByID(id);
    
    if (std::holds_alternative<DatabaseError>(getResult))
    {
        throw DatabaseException("Failed to retrieve updated editor");
    }
    
    return DtoMapper::ToResponse(std::get<TblEditor>(getResult));
}

bool EditorService::Delete(int64_t id)
{
    auto result = m_dao->Delete(id);
    
    if (std::holds_alternative<DatabaseError>(result))
    {
        DatabaseError error = std::get<DatabaseError>(result);
        if (error == DatabaseError::NotFound)
        {
            throw NotFoundException("Editor not found for deletion");
        }
        throw DatabaseException("Failed to delete editor");
    }
    
    return std::get<bool>(result);
}

std::vector<EditorResponseTo> EditorService::GetAll()
{
    auto result = m_dao->ReadAll();
    
    if (std::holds_alternative<DatabaseError>(result))
    {
        throw DatabaseException("Failed to retrieve all editors");
    }
    
    return DtoMapper::ToResponseList(std::get<std::vector<TblEditor>>(result));
}

}

File: 351002/Lashchenka/publisher/src/main/java/by/bsuir/distcomp/service/EditorService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.EditorRequestTo;
import by.bsuir.distcomp.dto.response.EditorResponseTo;
import by.bsuir.distcomp.entity.Editor;
import by.bsuir.distcomp.exception.DuplicateException;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.EditorMapper;
import by.bsuir.distcomp.model.EditorRole;
import by.bsuir.distcomp.repository.EditorRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class EditorService {

    private final EditorRepository editorRepository;
    private final EditorMapper editorMapper;
    private final PasswordEncoder passwordEncoder;

    public EditorService(EditorRepository editorRepository, EditorMapper editorMapper, PasswordEncoder passwordEncoder) {
        this.editorRepository = editorRepository;
        this.editorMapper = editorMapper;
        this.passwordEncoder = passwordEncoder;
    }

    @Caching(put = @CachePut(value = "editors", key = "#result.id"),
            evict = @CacheEvict(value = "editors", key = "'all'"))
    public EditorResponseTo create(EditorRequestTo dto) {
        if (editorRepository.existsByLogin(dto.getLogin())) {
            throw new DuplicateException("Editor with login '" + dto.getLogin() + "' already exists", 40301);
        }
        Editor entity = editorMapper.toEntity(dto);
        entity.setRole(EditorRole.CUSTOMER);
        entity.setPassword(encodePassword(dto.getPassword()));
        Editor saved = editorRepository.save(entity);
        return editorMapper.toResponseDto(saved);
    }

    @Cacheable(value = "editors", key = "#id")
    @Transactional(readOnly = true)
    public EditorResponseTo getById(Long id) {
        Editor entity = editorRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Editor with id " + id + " not found", 40401));
        return editorMapper.toResponseDto(entity);
    }

    @Cacheable(value = "editors", key = "'all'")
    @Transactional(readOnly = true)
    public List<EditorResponseTo> getAll() {
        return editorRepository.findAll().stream()
                .map(editorMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    @Caching(put = @CachePut(value = "editors", key = "#result.id"),
            evict = @CacheEvict(value = "editors", key = "'all'"))
    public EditorResponseTo update(EditorRequestTo dto) {
        Editor existing = editorRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Editor with id " + dto.getId() + " not found", 40402));
        if (editorRepository.existsByLoginAndIdNot(dto.getLogin(), dto.getId())) {
            throw new DuplicateException("Editor with login '" + dto.getLogin() + "' already exists", 40302);
        }
        editorMapper.updateEntityFromDto(dto, existing);
        if (dto.getPassword() != null && !dto.getPassword().isEmpty()) {
            existing.setPassword(encodePassword(dto.getPassword()));
        }
        Editor updated = editorRepository.save(existing);
        return editorMapper.toResponseDto(updated);
    }

    @Caching(evict = {
            @CacheEvict(value = "editors", key = "#id"),
            @CacheEvict(value = "editors", key = "'all'")
    })
    public void deleteById(Long id) {
        if (!editorRepository.existsById(id)) {
            throw new ResourceNotFoundException("Editor with id " + id + " not found", 40403);
        }
        editorRepository.deleteById(id);
    }

    private String encodePassword(String raw) {
        if (raw == null) {
            return null;
        }
        if (raw.startsWith("$2a$") || raw.startsWith("$2b$") || raw.startsWith("$2y$")) {
            return raw;
        }
        return passwordEncoder.encode(raw);
    }
}
